Codechange: Add font cache factory providers.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Instantiating `FontCache`s is a bit precarious, involving #defines depending on the platform / optionals being compiled, implementing the same function in different places.

Then there is the self-registering aspect. Each FontCache instantiated with `new` automatically adds itself to the FontCaches array, replacing the existing FontCache.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Switch to using the existing `ProviderManager` system, and using `FontCacheFactory`s to create the appropriate `FontCache`.

`FontCache` lifetime is now managed with `unique_ptr`, and the self-registering is removed.

Of course these still depend on #defines and build options, but they are now no longer spread out in various places, and implementing the provider and factory is standardised.

Built on #14448 and #14449

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/fontcache.h:
```diff
@@ -20,18 +21,23 @@ static const GlyphID SPRITE_GLYPH = 1U << 30;
-	FontCache *parent;                ///< The parent of this font cache.
-	const FontSize fs;                ///< The size of the font.
+	static std::array<std::unique_ptr<FontCache>, FS_END> caches; ///< All the font caches.
+	std::unique_ptr<FontCache>parent; ///< The parent of this font cache.
```
```suggestion
	std::unique_ptr<FontCache> parent; ///< The parent of this font cache.
```

rubidium42 commented on src/fontcache.h:
```diff
@@ -20,18 +21,23 @@ static const GlyphID SPRITE_GLYPH = 1U << 30;
 public:
-	FontCache(FontSize fs);
-	virtual ~FontCache();
+	virtual ~FontCache() {}
```
```suggestion
	virtual ~FontCache() = default;
```

rubidium42 commented on src/fontcache.cpp:
```diff
@@ -211,43 +220,55 @@ std::string GetFontCacheFontName(FontSize fs)
+ */
+/* static */ void FontCache::UninitializeFontCaches()
+{
+	std::ranges::generate(FontCache::caches, []() { return nullptr; });
```
This feels a bit weird; generating something frees things? Maybe something along the lines of the following is a conceptually more logical alternative:
`std::ranges::for_each(FontCache::caches, std::unique_ptr<FontCache>::reset);`?

rubidium42 commented on src/fontcache/freetypefontcache.cpp:
```diff
@@ -296,19 +206,116 @@ GlyphID FreeTypeFontCache::MapCharToGlyph(char32_t key, bool allow_fallback)
+
+	/**
+	 * Loads the freetype font.
+	 * First type to load the fontname as if it were a path. If that fails,
```
```suggestion
	 * First try to load the fontname as if it were a path. If that fails,
```
This was already wrong before this PR.

rubidium42 commented on src/fontcache/freetypefontcache.cpp:
```diff
@@ -296,19 +206,116 @@ GlyphID FreeTypeFontCache::MapCharToGlyph(char32_t key, bool allow_fallback)
+};
 
-#endif /* !defined(WITH_FONTCONFIG) */
+static FreeTypeFontCacheFactory s_freetype_fontcache_factory;
```
I'm only noticing this now, and did not with the sound loader and screenshot provider... but these shouldn't have the `s` prefix. I know that the blitter and drivers do use `s`, but those are local static variables. These are global variables.

rubidium42 commented on src/fontcache.cpp:
```diff
@@ -30,22 +28,38 @@
+			return true;
+		}
+	}
+	return false;
```
This might potentially be done with `std::ranges::any_of`. Something like 
```c++
return std::ranges::any_of(FontProviderManager::GetProviders(), 
        [&](provider) { return provider->FindFallbackFont(settings, language_iso_code, callback); }
````

PeterN committed:
Codechange: Use ProviderManager interface to register FontCache factories.

This removes use of #ifdefs to select the appropriate loader, and also replaces FontCache self-registration.

PeterN committed:
Codechange: Move fallback font detection to FontCacheFactory.

Provides a standard interface instead of relying on defines.
